<h3>Exercise 1<h2>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from scipy import stats
from scipy.stats import zscore
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
%load_ext numpy_html

In [3]:
%%html
<style>
table {margin-top: 16pt;}
td {font-family: Arial, Helvetica, sans-serif !important; font-weight: 600; white-space: normal !important;}
tr:nth-child(even) {background: #b0d9eb !important; color: black}
tr:nth-child(odd) {background: #8fb1bf !important; color: black}
</style>

In [4]:
# Read csv-file (loans.csv)
df = pd.read_csv('loans.csv')
pd.set_option('display.float_format', lambda x: '%.2f' % x)

# Drop column 'Customer ID'
df = df.drop('Customer ID', axis=1)

In [5]:
# Print the head of data, way 1
print(df.head())

                                Loan ID Loan Status  Current Loan Amount  \
0  14dd8831-6af5-400b-83ec-68e61888a048  Fully Paid            445412.00   
1  4771cc26-131a-45db-b5aa-537ea4ba5342  Fully Paid            262328.00   
2  4eed4e6a-aa2f-4c91-8651-ce984ee8fb26  Fully Paid          99999999.00   
3  77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a  Fully Paid            347666.00   
4  d4062e70-befa-4995-8643-a0de73938182  Fully Paid            176220.00   

         Term  Credit Score  Annual Income Years in current job  \
0  Short Term        709.00     1167493.00              8 years   
1  Short Term           NaN            NaN            10+ years   
2  Short Term        741.00     2231892.00              8 years   
3   Long Term        721.00      806949.00              3 years   
4  Short Term           NaN            NaN              5 years   

  Home Ownership             Purpose  Monthly Debt  Years of Credit History  \
0  Home Mortgage   Home Improvements       5214.74           

In [6]:
# Print the head of data, way 2 (I prefer this one)
df.head()

Loan ID Loan Status  Current Loan Amount  \
0  14dd8831-6af5-400b-83ec-68e61888a048  Fully Paid            445412.00   
1  4771cc26-131a-45db-b5aa-537ea4ba5342  Fully Paid            262328.00   
2  4eed4e6a-aa2f-4c91-8651-ce984ee8fb26  Fully Paid          99999999.00   
3  77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a  Fully Paid            347666.00   
4  d4062e70-befa-4995-8643-a0de73938182  Fully Paid            176220.00   

         Term  Credit Score  Annual Income Years in current job  \
0  Short Term        709.00     1167493.00              8 years   
1  Short Term           NaN            NaN            10+ years   
2  Short Term        741.00     2231892.00              8 years   
3   Long Term        721.00      806949.00              3 years   
4  Short Term           NaN            NaN              5 years   

  Home Ownership             Purpose  Monthly Debt  Years of Credit History  \
0  Home Mortgage   Home Improvements       5214.74                    17.20   
1  Home Mortgage  Debt Consolidation      33295.98                    21.10   
2       Own Home  Debt Consolidation      29200.53                    14.90   
3       Own Home  Debt Consolidation       8741.90                    12.00   
4           Rent  Debt Consolidation      20639.70                     6.10   

   Months since last delinquent  Number of Open Accounts  \
0                           NaN                     6.00   
1                          8.00                    35.00   
2                         29.00                    18.00   
3                           NaN                     9.00   
4                           NaN                    15.00   

   Number of Credit Problems  Current Credit Balance  Maximum Open Credit  \
0                       1.00               228190.00            416746.00   
1                       0.00               229976.00            850784.00   
2                       1.00               297996.00            750090.00   
3                       0.00               256329.00            386958.00   
4                       0.00               253460.00            427174.00   

   Bankruptcies  Tax Liens  
0          1.00       0.00  
1          0.00       0.00  
2          0.00       0.00  
3          0.00       0.00  
4          0.00       0.00

In [7]:
# Remove rows from data that have larger loan than  99999999
df = df[df["Current Loan Amount"] < 99999999].sort_values(by='Current Loan Amount', ascending=True)
df

Loan ID  Loan Status  Current Loan Amount  \
87741  7fd15008-5365-4385-bace-c093b1837555   Fully Paid             10802.00   
13338  20c4e560-e543-486a-a010-12bcdbc1b3ff   Fully Paid             11242.00   
19847  50118373-dd2c-402d-918b-a9e49d921eac   Fully Paid             15422.00   
79169  50118373-dd2c-402d-918b-a9e49d921eac   Fully Paid             15422.00   
79712  8e479ae7-12cc-4a76-bb92-ae6effa0b20c   Fully Paid             21098.00   
...                                     ...          ...                  ...   
94523  12092f92-171b-4d9b-89ae-82ca88bfc26b   Fully Paid            789184.00   
55736  8d59aa00-d111-4439-9b32-82cc5ab5b940   Fully Paid            789184.00   
42195  b7923e12-395b-49d6-97ee-ca0ed25b65c2   Fully Paid            789250.00   
82002  770ca5a2-3b53-4c8c-a238-55cf72f66b82  Charged Off            789250.00   
21017  4fed402f-4298-4f5c-8d30-05b1f1819c00   Fully Paid            789250.00   

             Term  Credit Score  Annual Income Years in current job  \
87741  Short Term           NaN            NaN             < 1 year   
13338  Short Term        747.00      349391.00             < 1 year   
19847  Short Term        718.00      361760.00             < 1 year   
79169  Short Term        718.00      361760.00             < 1 year   
79712  Short Term           NaN            NaN               1 year   
...           ...           ...            ...                  ...   
94523  Short Term        744.00     2142041.00              2 years   
55736   Long Term        716.00     1596798.00              3 years   
42195  Short Term        721.00     8054480.00              6 years   
82002  Short Term           NaN            NaN              2 years   
21017   Long Term        672.00     1752750.00            10+ years   

      Home Ownership             Purpose  Monthly Debt  \
87741   HaveMortgage         Take a Trip        373.73   
13338           Rent               other       1243.17   
19847           Rent  Debt Consolidation       3738.06   
79169           Rent  Debt Consolidation       3738.06   
79712           Rent               other       8231.56   
...              ...                 ...           ...   
94523  Home Mortgage  Debt Consolidation      15351.24   
55736  Home Mortgage  Debt Consolidation      44710.23   
42195  Home Mortgage  Debt Consolidation     114105.07   
82002  Home Mortgage  Debt Consolidation      17267.96   
21017           Rent  Debt Consolidation      29212.50   

       Years of Credit History  Months since last delinquent  \
87741                    27.40                          8.00   
13338                    12.70                          0.00   
19847                    11.60                           NaN   
79169                    11.60                           NaN   
79712                    20.60                          0.00   
...                        ...                           ...   
94523                    21.20                           NaN   
55736                    28.50                           NaN   
42195                    31.40                          2.00   
82002                    21.00                           NaN   
21017                     8.50                           NaN   

       Number of Open Accounts  Number of Credit Problems  \
87741                     3.00                       0.00   
13338                     4.00                       0.00   
19847                     4.00                       0.00   
79169                     4.00                       0.00   
79712                     8.00                       0.00   
...                        ...                        ...   
94523                    10.00                       0.00   
55736                    15.00                       0.00   
42195                    21.00                       0.00   
82002                     8.00                       0.00   
21017                    11.00                       0.00   

       Curr

In [8]:
# Use reindex for more understandable view
df = df.reset_index(drop=True)
df

Loan ID  Loan Status  Current Loan Amount  \
0      7fd15008-5365-4385-bace-c093b1837555   Fully Paid             10802.00   
1      20c4e560-e543-486a-a010-12bcdbc1b3ff   Fully Paid             11242.00   
2      50118373-dd2c-402d-918b-a9e49d921eac   Fully Paid             15422.00   
3      50118373-dd2c-402d-918b-a9e49d921eac   Fully Paid             15422.00   
4      8e479ae7-12cc-4a76-bb92-ae6effa0b20c   Fully Paid             21098.00   
...                                     ...          ...                  ...   
88511  12092f92-171b-4d9b-89ae-82ca88bfc26b   Fully Paid            789184.00   
88512  8d59aa00-d111-4439-9b32-82cc5ab5b940   Fully Paid            789184.00   
88513  b7923e12-395b-49d6-97ee-ca0ed25b65c2   Fully Paid            789250.00   
88514  770ca5a2-3b53-4c8c-a238-55cf72f66b82  Charged Off            789250.00   
88515  4fed402f-4298-4f5c-8d30-05b1f1819c00   Fully Paid            789250.00   

             Term  Credit Score  Annual Income Years in current job  \
0      Short Term           NaN            NaN             < 1 year   
1      Short Term        747.00      349391.00             < 1 year   
2      Short Term        718.00      361760.00             < 1 year   
3      Short Term        718.00      361760.00             < 1 year   
4      Short Term           NaN            NaN               1 year   
...           ...           ...            ...                  ...   
88511  Short Term        744.00     2142041.00              2 years   
88512   Long Term        716.00     1596798.00              3 years   
88513  Short Term        721.00     8054480.00              6 years   
88514  Short Term           NaN            NaN              2 years   
88515   Long Term        672.00     1752750.00            10+ years   

      Home Ownership             Purpose  Monthly Debt  \
0       HaveMortgage         Take a Trip        373.73   
1               Rent               other       1243.17   
2               Rent  Debt Consolidation       3738.06   
3               Rent  Debt Consolidation       3738.06   
4               Rent               other       8231.56   
...              ...                 ...           ...   
88511  Home Mortgage  Debt Consolidation      15351.24   
88512  Home Mortgage  Debt Consolidation      44710.23   
88513  Home Mortgage  Debt Consolidation     114105.07   
88514  Home Mortgage  Debt Consolidation      17267.96   
88515           Rent  Debt Consolidation      29212.50   

       Years of Credit History  Months since last delinquent  \
0                        27.40                          8.00   
1                        12.70                          0.00   
2                        11.60                           NaN   
3                        11.60                           NaN   
4                        20.60                          0.00   
...                        ...                           ...   
88511                    21.20                           NaN   
88512                    28.50                           NaN   
88513                    31.40                          2.00   
88514                    21.00                           NaN   
88515                     8.50                           NaN   

       Number of Open Accounts  Number of Credit Problems  \
0                         3.00                       0.00   
1                         4.00                       0.00   
2                         4.00                       0.00   
3                         4.00                       0.00   
4                         8.00                       0.00   
...                        ...                        ...   
88511                    10.00                       0.00   
88512                    15.00                       0.00   
88513                    21.00                       0.00   
88514                     8.00                       0.00   
88515                    11.00                       0.00   

       Curr

In [9]:
# Check if there is any na (missing values) within 'Annual Income Column
df['Annual Income'].isna().sum()

19154

In [10]:
# Count average 'Annual Income'
avg_annual_income = df['Annual Income'].mean()

# Use imputation method to fill NaN (not a number) values in 'Annual Income' column
df['Annual Income'].fillna(avg_annual_income, inplace=True)
df = df[df['Annual Income'].notna()]
df['Annual Income']

C:\Users\Archoste\AppData\Local\Temp\ipykernel_16980\3695469613.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Annual Income'].fillna(avg_annual_income, inplace=True)


0       1373245.98
1        349391.00
2        361760.00
3        361760.00
4       1373245.98
           ...    
88511   2142041.00
88512   1596798.00
88513   8054480.00
88514   1373245.98
88515   1752750.00
Name: Annual Income, Length: 88516, dtype: float64

In [11]:
# Check if imputation method worked
df['Annual Income'].isna().sum()

0

In [12]:
# Get the average 'Current Loan Amount'
avg_loan_amt = df['Current Loan Amount'].mean()
print(f'Average loan amount: {avg_loan_amt}$')

Average loan amount: 312313.59788060916$


In [13]:
# Get the highest & lowest 'Annual Income from the dataset
highest_annual_income = df['Annual Income'].max()
lowest_annual_income = df['Annual Income'].min()
print(f'Highest Annual Income: {highest_annual_income}$')
print(f'Lowest Annual Income: {lowest_annual_income}$')

Highest Annual Income: 165557393.0$
Lowest Annual Income: 76627.0$


In [14]:
# Get the Home Ownership value of the Loan ID = 'bbf87a87-22cd-4d10-bd9b-7a9cc1b6e59d'

home_ownership_value= df[df['Loan ID'] == 'bbf87a87-22cd-4d10-bd9b-7a9cc1b6e59d']['Home Ownership'].iloc[0]
print(f'Home Ownership State of asked Loan ID:', home_ownership_value)

Home Ownership State of asked Loan ID: Own Home


In [15]:
# Create a newfield called Actual Annual Income into your dataset with formula below,
# The Actual Annual Income follow this formula: Annual Income – 12 * Monthly Debt
df['Actual Annual Income'] = df['Annual Income'] - 12 * df['Monthly Debt']
df['Actual Annual Income']

0       1368761.22
1        334472.96
2        316903.28
3        316903.28
4       1274467.26
           ...    
88511   1957826.12
88512   1060275.24
88513   6685219.16
88514   1166030.46
88515   1402200.00
Name: Actual Annual Income, Length: 88516, dtype: float64

In [16]:
# Get the Actual Annual Income of loan ID: "76fa89b9-e6a8-49af-afa1-8151315aba8e"]
actual_annual_income = df[df["Loan ID"] == "76fa89b9-e6a8-49af-afa1-8151315aba8e"]["Actual Annual Income"].iloc[0]
print(f"Actual Annual Income: {actual_annual_income}$")

Actual Annual Income: 674234.76$


In [17]:
# Get the loan amount of smallest Actual Annual Income
min_actual_annual_income_id = df[df['Actual Annual Income'] == df['Actual Annual Income'].min()]['Loan ID'].iloc[0]
print(f'Loan ID with the smallest Actual Annual Income:', min_actual_annual_income_id)


Loan ID with the smallest Actual Annual Income: 0a84b8ab-ed2f-474f-a3a9-8bf5bb90c31a


In [18]:
# Get the amount of long term loans, way 1
long_term_loans = df[df['Term'] == 'Long Term'].shape[0]
print(f'Amount of long term loans:', long_term_loans)


Amount of long term loans: 25475


In [19]:
# Use function value_counts to get amount of short & long term loans
# This is way 2, this way you get both
df['Term'].value_counts()

Term
Short Term    63041
Long Term     25475
Name: count, dtype: int64

In [20]:
# Get the amount of loaners who have more than 1 bankruptcy
several_bankrupcies = df[df['Bankruptcies'] > 1].shape[0]
print(f'Loaners with several bankruptcies:', several_bankrupcies)

Loaners with several bankruptcies: 476


In [21]:
# Get number of short term loans for the Home improvements (Hil, Home Improvement Loan)
short_term_HIL = df[(df['Term']=='Short Term') & (df['Purpose'] == 'Home Improvements')].shape[0]
print(f'Number of short term loan for the home improvements:', short_term_HIL)

Number of short term loan for the home improvements: 3490


In [22]:
# Get the number of unique loan purposes
unique_loan_purposes = df['Purpose'].unique()
number_unique_loan_purposes = df['Purpose'].nunique()
print(f'Unique loan purposes:', unique_loan_purposes)
print(f'Number of unique loan purposes:', number_unique_loan_purposes)

Unique loan purposes: ['Take a Trip' 'other' 'Debt Consolidation' 'Medical Bills' 'Other'
 'Home Improvements' 'major_purchase' 'Buy a Car' 'vacation' 'moving'
 'wedding' 'renewable_energy' 'Business Loan' 'Buy House'
 'Educational Expenses' 'small_business']
Number of unique loan purposes: 16


In [23]:
# Get the three most common loan purposes
three_common_loan_purposes = df['Purpose'].value_counts()
three_common_loan_purposes.head(3)

Purpose
Debt Consolidation    69506
other                  5409
Home Improvements      5124
Name: count, dtype: int64

In [24]:
#Correlation problem using .corr()
correlation1 = df['Annual Income'].corr(df['Number of Open Accounts'])
correlation2 = df['Number of Credit Problems'].corr(df['Bankruptcies'])
print(f'Here is the correlation between Annual Income and Number of Open Accounts:', correlation1)
print(f'Here is the correlation between Number of credit problems and Bankruptcies:', correlation2)

Here is the correlation between Annual Income and Number of Open Accounts: 0.12705769402525524
Here is the correlation between Number of credit problems and Bankruptcies: 0.7507485199510441


This informs us that correlation between Annual Income & Number of Loans columns is quite low. This could be caused by higher income inviduals having several accounts open at this bank.

Other hand, correlation between Number of Credit Problems and Bankruptcies is relatively high, this could be caused by people who have income problems being more prone for filing more bankruptcies.

<h3>Exercise 2<h3>

In [25]:
# Read csv-file (purchase csv)
df = pd.read_csv("purchases.csv")

In [26]:
# Print the head of the data
df.head()

Creation Date Purchase Date Fiscal Year LPA Number Purchase Order Number  \
0    09/05/2012    08/31/2012   2012-2013        NaN            4500149558   
1    10/18/2012    10/18/2012   2012-2013        NaN            4500156192   
2    10/17/2012    10/17/2012   2012-2013        NaN            4500156124   
3    10/30/2012    10/24/2012   2012-2013        NaN              S2556056   
4    11/14/2012    11/14/2012   2012-2013        NaN            4500159228   

   Requisition Number Acquisition Type    Acquisition Method  \
0                 NaN     NON-IT Goods    Formal Competitive   
1                 NaN     NON-IT Goods  Informal Competitive   
2                 NaN     NON-IT Goods   Fair and Reasonable   
3                 NaN     NON-IT Goods  Informal Competitive   
4                 NaN     NON-IT Goods  Informal Competitive   

                                 Department Name  Supplier Code  ...  \
0  Corrections and Rehabilitation, Department of       48199.00  ...   
1  Corrections and Rehabilitation, Department of     1327540.00  ...   
2              Correctional Health Care Services       45803.00  ...   
3              Parks & Recreation, Department of     1746707.00  ...   
4  Corrections and Rehabilitation, Department of     1065902.00  ...   

  Classification Codes Normalized UNSPSC   Commodity Title       Class  \
0             50405625       50405625.00  Jalapeno peppers 50405600.00   
1             50301541       50301541.00    Ida red apples 50301500.00   
2             55101506       55101506.00         Magazines 55101500.00   
3             10121505       10121505.00               Hay 10121500.00   
4             50201706       50201706.00            Coffee 50201700.00   

            Class Title      Family      Family Title     Segment  \
0               Peppers 50400000.00  Fresh vegetables 50000000.00   
1                Apples 50300000.00      Fresh fruits 50000000.00   
2  Printed publications 55100000.00     Printed media 55000000.00   
3        Livestock feed 10120000.00       Animal feed 10000000.00   
4        Coffee and tea 50200000.00         Beverages 50000000.00   

                                       Segment Title  \
0                 Food Beverage and Tobacco Products   
1                 Food Beverage and Tobacco Products   
2                                 Published Products   
3  Live Plant and Animal Material and Accessories...   
4                 Food Beverage and Tobacco Products   

                          Location  
0  93706\n(36.675079, -119.865393)  
1  91360\n(34.210392, -118.874313)  
2  95827\n(38.563097, -121.328511)  
3                              NaN  
4          95696\n(38.43, -122.02)  

[5 rows x 29 columns]

In [27]:
# Get the total price sum of Purchase Order Number (PON) 018H2015
pon_price = df[df['Purchase Order Number'] == '018H2015']['Total Price'].sum()
total_price = round(pon_price, 2)
print(f'The total price for the purchase order number: {total_price}$')

The total price for the purchase order number: 1245.04$


In [28]:
# Get the name and description of the item with PON 3176273
item_information = df[df['Purchase Order Number'] == '3176273'][['Item Name', 'Item Description']].iloc[0]
print(f'The name & description of the item:\n{item_information}')

The name & description of the item:
Item Name               PC Desktop
Item Description    HP Prodesk 600
Name: 103, dtype: object


In [29]:
# Get the rows of purchased data happened during the year 2013
purch_data_2013 = df[df['Purchase Date'].str.contains('2013')].shape[0]
print(f'The purchase data in 2013 is: {purch_data_2013}')

The purchase data in 2013 is: 34024


In [30]:
# Get the five most common departments from the data
five_common_deps = df['Department Name'].value_counts().head(5)
print('Five most common Departments:')
print(f"{five_common_deps}")


Five most common Departments:
Department Name
Corrections and Rehabilitation, Department of    10891
Correctional Health Care Services                 6593
Water Resources, Department of                    4582
Forestry and Fire Protection, Department of       4526
State Hospitals, Department of                    4281
Name: count, dtype: int64


In [31]:
# Get three the most money using Departments
sorted_by_total_sum = df.groupby('Department Name')['Total Price'].sum()
sorted_by_total_sum.head(3)

Department Name
Administrative Law, Office of            50000.00
African American Museum, California     103122.44
Aging, Department of                  82834987.10
Name: Total Price, dtype: float64

In [32]:
# Sort the data by Department Name
sorted_by_name = df.sort_values('Department Name')
sorted_by_name


Creation Date Purchase Date Fiscal Year LPA Number  \
6        11/30/2012    06/15/2012   2012-2013        NaN   
519      11/30/2012    06/15/2012   2012-2013        NaN   
20018    11/30/2012    06/15/2012   2012-2013        NaN   
31381    02/21/2013    03/20/2013   2012-2013        NaN   
27820    01/03/2013    02/14/2011   2012-2013        NaN   
...             ...           ...         ...        ...   
24826    12/17/2012    12/15/2012   2012-2013        NaN   
49068    06/13/2013    06/12/2013   2012-2013        NaN   
49069    03/26/2014    03/25/2014   2013-2014        NaN   
49057    06/13/2013    06/12/2013   2012-2013        NaN   
44469    05/07/2013    05/01/2013   2012-2013        NaN   

      Purchase Order Number  Requisition Number Acquisition Type  \
6                   11OAL07                 NaN     NON-IT Goods   
519                 11OAL09                 NaN     NON-IT Goods   
20018               11OAL10                 NaN     NON-IT Goods   
31381              11-12201                 NaN  NON-IT Services   
27820              11-12175                 NaN  NON-IT Services   
...                     ...                 ...              ...   
24826            4500192615                 NaN     NON-IT Goods   
49068            4500199013                 NaN         IT Goods   
49069            4500208992                 NaN     NON-IT Goods   
49057            4500199013                 NaN         IT Goods   
44469            4500197248                 NaN     NON-IT Goods   

         Acquisition Method                      Department Name  \
6                       NCB        Administrative Law, Office of   
519                     NCB        Administrative Law, Office of   
20018                   NCB        Administrative Law, Office of   
31381  Informal Competitive  African American Museum, California   
27820  Informal Competitive  African American Museum, California   
...                     ...                                  ...   
24826    Formal Competitive       Water Resources, Department of   
49068  Informal Competitive       Water Resources, Department of   
49069        State Programs       Water Resources, Department of   
49057  Informal Competitive       Water Resources, Department of   
44469        State Programs       Water Resources, Department of   

       Supplier Code  ... Classification Codes Normalized UNSPSC  \
6         1001493.00  ...             55101524       55101524.00   
519       1657720.00  ...             55101524       55101524.00   
20018     1657720.00  ...             55101524       55101524.00   
31381        7341.00  ...             82101801       82101801.00   
27820     1749414.00  ...             78141501       78141501.00   
...              ...  ...                  ...               ...   
24826       16299.00  ...   27131603\n40141605       27131603.00   
49068     1016373.00  ...             43232605       43232605.00   
49069     1743406.00  ...             15101506       15101506.00   
49057     1016373.00  ...             81112217       81112217.00   
44469     1743406.00  ...             78180301       78180301.00   

                              Commodity Title       Class  \
6                             Reference books 55101500.00   
519                           Reference books 55101500.00   
20018                         Reference books 55101500.00   
31381           Advertising campaign services 82101800.00   
27820             Freight forwarders services 78141500.00   
...                                       ...         ...   
24826                          Air regulators 27131600.00   
49068       Analytical or scientific software 43232600.00   
49069                      Gasoline or Petrol 15101500.00   
49057  Industry specific software maintenance 81112200.00   
44469                    Fuelling of vehicles 78180300.00   

                                             Class Title      Family  \
6                            

In [33]:
# Replace the dollar sign of Unit Price column, and turning it into a float.
df['Unit Price'] = df['Unit Price'].str.replace('$', " ").astype(float)
df['Unit Price']

0           0.60
1       12587.05
2         199.00
3           4.95
4           1.00
          ...   
62652     100.50
62653     245.00
62654     245.00
62655   20475.00
62656    1181.44
Name: Unit Price, Length: 62657, dtype: float64

<h3> Extra tasks:<h3>

In [34]:
# Calculate the total amount of purchases > 50 000 dollars
purchs_over_50k = (df['Total Price'] > 50000).sum()
print(f'There has been: {purchs_over_50k} purchases over 50 000$.')



There has been: 6761 purchases over 50 000$.


In [35]:
# Provide prove of your calculations
check_up = df['Total Price'] > 50000
check_up.value_counts()

Total Price
False    55896
True      6761
Name: count, dtype: int64

In [36]:
# Define IT Departments
it_departments = ['IT Goods', 'IT Services', 'IT Telecommunications']

# Filter purchases related to IT departments
it_related_purchases = df[df['Acquisition Type'].isin(it_departments) & (df['Total Price'] > 50000)]
it_related_purchases = it_related_purchases['Acquisition Type'].count()
print(f'There has been: {it_related_purchases} purchases valued over 50 000$.')

There has been: 1153 purchases valued over 50 000$.


In [37]:
# Find rows that have purchase Total Price 0 < less
purchs_under_zero = df[df['Total Price'] <= 0]
purchs_under_zero


Creation Date Purchase Date Fiscal Year  LPA Number  \
254      09/19/2013    09/19/2013   2013-2014         NaN   
278      03/08/2013    03/08/2013   2012-2013         NaN   
302      12/14/2012    12/12/2012   2012-2013         NaN   
354      03/27/2014    03/24/2014   2013-2014         NaN   
360      06/07/2014    06/05/2014   2013-2014         NaN   
...             ...           ...         ...         ...   
62427    08/08/2013    06/21/2013   2013-2014  7-09-36-04   
62439    08/12/2013    10/01/2012   2013-2014         NaN   
62476    08/13/2013    05/22/2013   2013-2014         NaN   
62481    08/14/2013    08/12/2013   2013-2014  1-12-65-50   
62600    08/19/2013    08/19/2013   2013-2014         NaN   

      Purchase Order Number  Requisition Number Acquisition Type  \
254              4500199875                 NaN     NON-IT Goods   
278              4500172951                 NaN     NON-IT Goods   
302               SA12 9009                 NaN  NON-IT Services   
354                   X1762                 NaN     NON-IT Goods   
360                   31317                 NaN     NON-IT Goods   
...                     ...                 ...              ...   
62427              SM12-733                 NaN      IT Services   
62439              CN120218                 NaN  NON-IT Services   
62476              13-I0002                 NaN         IT Goods   
62481              PR350050                 NaN     NON-IT Goods   
62600            5600003989                 NaN  NON-IT Services   

                                Acquisition Method  \
254                                            NCB   
278                           Informal Competitive   
302                            Fair and Reasonable   
354                           Informal Competitive   
360                                            NCB   
...                                            ...   
62427                                    WSCA/Coop   
62439  Services are specifically exempt by statute   
62476                          Fair and Reasonable   
62481                           Statewide Contract   
62600  Services are specifically exempt by statute   

                                     Department Name  Supplier Code  ...  \
254    Corrections and Rehabilitation, Department of     1065902.00  ...   
278                Correctional Health Care Services       11350.00  ...   
302                              Air Resources Board     1740063.00  ...   
354                           Justice, Department of     1770456.00  ...   
360                   Social Services, Department of     1755837.00  ...   
...                                              ...            ...  ...   
62427                         Student Aid Commission     1232200.00  ...   
62439                       Education, Department of     1743598.00  ...   
62476            Pesticide Regulation, Department of       29828.00  ...   
62481          Developmental Services, Department of     1010439.00  ...   
62600  Corrections and Rehabilitation, Department of     1012530.00  ...   

      Classification Codes Normalized UNSPSC  \
254               56112111       56112111.00   
278               50121537       50121537.00   
302               82101504       82101504.00   
354               84111506       84111506.00   
360               56101532       56101532.00   
...                    ...               ...   
62427             81112306       81112306.00   
62439             86141501       86141501.00   
62476             81112220       81112220.00   
62481             41116004       41116004.00   
62600   85122103\n85141701       85122103.00   

                                   Commodity Title       Class  \
254                                  Booster seats 56112100.00   
278                                    Frozen fish 50121500.00   
302                          Newspaper advertising 82101500.00   
354                               Billing services 84

In [38]:
# Create DataFrame without rows that have Total Price <= 0 or less
df = df.drop(purchs_under_zero)

KeyError: "['Creation Date', 'Purchase Date', 'Fiscal Year', 'LPA Number', 'Purchase Order Number', 'Requisition Number', 'Acquisition Type', 'Acquisition Method', 'Department Name', 'Supplier Code', 'Supplier Name', 'Supplier Qualifications', 'Supplier Zip Code', 'CalCard', 'Item Name', 'Item Description', 'Quantity', 'Unit Price', 'Total Price', 'Classification Codes', 'Normalized UNSPSC', 'Commodity Title', 'Class', 'Class Title', 'Family', 'Family Title', 'Segment', 'Segment Title', 'Location'] not found in axis"

In [ ]:
# Check that purchaces <= 0 are deleted
purchs_under_zero.sum()

Series([], Name: count, dtype: int64)

In [ ]:
# Use groupby function twice to group purchases data by Acquisition Type, and then calculating the result first by sum() and then by mean()
double_grouping = df.groupby('Acquisition Type')['Total Price'].agg(['sum', 'mean'])

# Group by sum
double_grouping.sort_values('sum', ascending=True)

sum       mean
Acquisition Type                               
IT Telecommunications      307903.39   51317.23
IT Goods                213676581.04   28953.47
NON-IT Goods           1034291622.78   26087.51
IT Services            1156983637.17  682989.16
NON-IT Services       30564202787.91 2481062.00

In [ ]:
# Group by sum
double_grouping.sort_values('mean', ascending=True)

sum       mean
Acquisition Type                               
NON-IT Goods           1034291622.78   26087.51
IT Goods                213676581.04   28953.47
IT Telecommunications      307903.39   51317.23
IT Services            1156983637.17  682989.16
NON-IT Services       30564202787.91 2481062.00

<h3>Here we can see that largest sum and mean values are owned by NON-IT Services, after that 2nd most largest sums own IT Services and 3rd largest value is owned by NON-IT GOODS. IT SERVICES own 2nd most largest mean and It Telecommications comes as 3rd.<h3>

<h2>Exercise 3: Data from salaries in India CSV</h2>

In [ ]:
# Read the csv-file(data_salaries_india.csv)
df = pd.read_csv('data_salaries_india.csv')

In [ ]:
# Get the columns names
df.columns

Index(['Company Name', 'Job Title', 'Role', 'Salaries Reported', 'Location',
       'Salary'],
      dtype='object')

In [ ]:
#Convert all salaries into yearly salary
def yearly_wage(row):
    period = row['Salary'][-2:]
    number = int(''.join(filter(str.isdigit, row['Salary'])))
    if period == 'hr':
        number = int(number * 2117.01)
    elif period == 'mo':
        # months to year
        number = int(number * 12)
    return number

In [ ]:
# Remove the Indian rupees sign
df['Salary'] = df['Salary'].str.replace('₹', '' '')
df['Salary']

0       11,71,687/yr
1       32,57,305/yr
2       58,11,205/yr
3       11,80,965/yr
4        6,02,930/yr
            ...     
5122     5,13,690/yr
5123    13,09,535/yr
5124    44,99,222/yr
5125    46,12,711/yr
5126    17,05,561/yr
Name: Salary, Length: 5127, dtype: object

In [ ]:
# Remove characters that aren't numbers & convert to float
df['Salary'] = df['Salary'].str.replace('[^\d.]', '', regex=True).astype(float)

In [ ]:
# Print the head the dataframe
df.head()

Company Name                  Job Title        Role  \
0                 Axis Bank             Data Scientist  Specialist   
1  Globex Digital Solutions  Machine Learning Engineer     Manager   
2        Tempworks Services             Data Scientist     Manager   
3                BlueOptima  Machine Learning Engineer  Specialist   
4         Dell Technologies               Data Analyst  Specialist   

   Salaries Reported   Location     Salary  
0                  3  Bangalore 1171687.00  
1                  1  Hyderabad 3257305.00  
2                  1  Bangalore 5811205.00  
3                  2  Bangalore 1180965.00  
4                  5  New Delhi  602930.00

In [ ]:
# Check for any missing values
missing_values = df.isna().sum()
print('Number of missing values in each column:', missing_values)

Number of missing values in each column: Company Name         4
Job Title            0
Role                 0
Salaries Reported    0
Location             0
Salary               0
dtype: int64


In [ ]:
# Remove rows with missing 'Company Name' values
df = df.dropna(subset=['Company Name'])
df

Company Name                  Job Title        Role  \
0                    Axis Bank             Data Scientist  Specialist   
1     Globex Digital Solutions  Machine Learning Engineer     Manager   
2           Tempworks Services             Data Scientist     Manager   
3                   BlueOptima  Machine Learning Engineer  Specialist   
4            Dell Technologies               Data Analyst  Specialist   
...                        ...                        ...         ...   
5122                   Rejolut  Machine Learning Engineer  Specialist   
5123     Bartleby Technologies  Machine Learning Engineer  Specialist   
5124                Clairvolex             Data Scientist     Manager   
5125       Piramal Enterprises               Data Analyst     Manager   
5126                  Unilever             Data Scientist  Specialist   

      Salaries Reported   Location     Salary  
0                     3  Bangalore 1171687.00  
1                     1  Hyderabad 3257305.00  
2                     1  Bangalore 5811205.00  
3                     2  Bangalore 1180965.00  
4                     5  New Delhi  602930.00  
...                 ...        ...        ...  
5122                  1     Mumbai  513690.00  
5123                  1  Bangalore 1309535.00  
5124                  1  Bangalore 4499222.00  
5125                  1     Mumbai 4612711.00  
5126                  2  Bangalore 1705561.00  

[5123 rows x 6 columns]

In [ ]:
# Find the most common values for job titles, company names, locations and salaries
job_title_counts = df['Job Title'].value_counts().head(5)
print(job_title_counts)

company_counts = df['Company Name'].value_counts().head(5)
print(company_counts)

location_counts = df['Location'].value_counts().head(5)
print(location_counts)

salary_counts = df["Salary"].value_counts().head(5)
print(salary_counts)

Job Title
Data Scientist               2165
Data Analyst                 1166
Data Engineer                 947
Machine Learning Engineer     683
Data Science                   54
Name: count, dtype: int64
Company Name
Tata Consultancy Services    51
Accenture                    39
Amazon                       35
Fresher                      31
First Student                30
Name: count, dtype: int64
Location
Bangalore    1858
Pune          962
Hyderabad     786
New Delhi     781
Mumbai        736
Name: count, dtype: int64
Salary
1200000.00    31
600000.00     29
1000000.00    24
500000.00     22
300000.00     20
Name: count, dtype: int64


In [ ]:
# Combine similar job titles
job_title_map = {
    'Machine Learning Data Associate': 'Machine Learning Associate',
    'Machine Learning Associate': 'Machine Learning Associate',
}
df['Job Title'] = df['Job Title'].map(job_title_map).fillna(df['Job Title'])


In [ ]:
# Analyze distribution to check the balance of the data
job_title_bal = len(job_title_counts) / len(df['Job Title'].unique())
company_bal = len(company_counts) / len(df['Company Name'].unique())
location_bal = len(location_counts) / len(df['Location'].unique())

# Display the information to the user
print(job_title_bal)
print(company_bal)
print(location_bal)

0.18518518518518517
0.0019817677368212444
1.0


In [ ]:
# Calculate observations  for each category, and compare them to see if the data is balanced or imbalanced
job_title_counts = df['Job Title'].value_counts()
company_counts = df['Company Name'].value_counts()
location_counts = df['Location'].value_counts()
salary_counts = df['Salary'].value_counts()

if job_title_counts.std() < 0.1 and company_counts.std() < 0.1 and location_counts.std() < 0.1 and salary_counts.std() < 0.1:
    print('The data is balanced.')
else:
    print('The data is unbalanced.')

The data is unbalanced.


<h3>Here in this comparison, we can see that data is unbalanced due to the deviation between the data being huge.<h3>

In [ ]:
# Calculate Z score for the Salary column
df['Salary_zscore'] = (df['Salary'] - df['Salary'].mean()) / df['Salary'].std()
df['Salary_zscore']


0      -0.03
1       1.11
2       2.52
3      -0.03
4      -0.35
        ... 
5122   -0.40
5123    0.04
5124    1.80
5125    1.86
5126    0.26
Name: Salary_zscore, Length: 5123, dtype: float64

In [ ]:
# Salary greater than z score of 3
df[df['Salary_zscore'] > 3]

Company Name                  Job Title  \
147                                Shell             Data Scientist   
228        Endurance International Group              Data Engineer   
385                            EMPLOYERS             Data Scientist   
399                              Kabbage             Data Scientist   
407                        Multi Recruit             Data Scientist   
425                     Johnson Controls             Data Scientist   
463                                Nones             Data Scientist   
892                                 Meta             Data Scientist   
945                            Kiwi labs             Data Scientist   
1106                          Freelancer             Data Scientist   
1184                     ABC Consultants             Data Scientist   
1239                         Wells Fargo              Data Engineer   
1245                                 ABC             Data Scientist   
1268       Batlivala & Karani Securities             Data Scientist   
1504                              Zomato              Data Engineer   
1620                                 TCS             Data Scientist   
1622                        Interactions  Machine Learning Engineer   
1644                       First Student  Machine Learning Engineer   
1668                  Wipro Technologies  Machine Learning Engineer   
1834                             Globant              Data Engineer   
1962                             Bizongo              Data Engineer   
2024                     Colors Software  Machine Learning Engineer   
2154                          Concentrix             Data Scientist   
2279                     Artis signature             Data Scientist   
2316                           Hashworks             Data Scientist   
2413                Orbinet Technologies             Data Scientist   
2497    Hewlett Packard Enterprise | HPE             Data Scientist   
2598                                 XYZ             Data Scientist   
2602                              Toptal             Data Scientist   
2647                             Birdeye             Data Scientist   
2652                              ABCDEF             Data Scientist   
2730                             Red Hat             Data Scientist   
2737                       Priceline.com              Data Engineer   
2866                              NVIDIA             Data Scientist   
2982                    Société Générale             Data Scientist   
3020                          ExxonMobil             Data Scientist   
3111                   Greenlight Planet  Machine Learning Engineer   
3181                  DreamOrbit Softech  Machine Learning Engineer   
3203                     Qwerty Concepts             Data Scientist   
3324                                Hike  Machine Learning Engineer   
3367                              Google             Data Scientist   
3623                          Freshworks             Data Scientist   
3672                    Juniper Networks             Data Scientist   
3725                  Super Highway Labs              Data Engineer   
3894                              Amazon          National Director   
4000                            L Brands             Data Scientist   
4047     Alfonso & Associates Consulting             Data Scientist   
4098                  Ritwik IT Services             Data Scientist   
4224  Zee Entertainment Enterprises Ltd.             Data Scientist   
4269                                Citi             Data Scientist   
4344                  Persistent Systems             Data Scientist   
4352                            i4u Labs             Data Scientist   
4541                        Thoughtworks             Data Scientist   
4626                  Aligned Automation      Senior Data Scientist   
4689                     ICS Consultancy  Machine Learning Engineer   
4750                          ICICI Bank            

In [ ]:
# Salary less than z score of -3.
df[df['Salary_zscore'] > -3]

Company Name                  Job Title        Role  \
0                    Axis Bank             Data Scientist  Specialist   
1     Globex Digital Solutions  Machine Learning Engineer     Manager   
2           Tempworks Services             Data Scientist     Manager   
3                   BlueOptima  Machine Learning Engineer  Specialist   
4            Dell Technologies               Data Analyst  Specialist   
...                        ...                        ...         ...   
5122                   Rejolut  Machine Learning Engineer  Specialist   
5123     Bartleby Technologies  Machine Learning Engineer  Specialist   
5124                Clairvolex             Data Scientist     Manager   
5125       Piramal Enterprises               Data Analyst     Manager   
5126                  Unilever             Data Scientist  Specialist   

      Salaries Reported   Location     Salary  Salary_zscore  
0                     3  Bangalore 1171687.00          -0.03  
1                     1  Hyderabad 3257305.00           1.11  
2                     1  Bangalore 5811205.00           2.52  
3                     2  Bangalore 1180965.00          -0.03  
4                     5  New Delhi  602930.00          -0.35  
...                 ...        ...        ...            ...  
5122                  1     Mumbai  513690.00          -0.40  
5123                  1  Bangalore 1309535.00           0.04  
5124                  1  Bangalore 4499222.00           1.80  
5125                  1     Mumbai 4612711.00           1.86  
5126                  2  Bangalore 1705561.00           0.26  

[5123 rows x 7 columns]

In [ ]:
# Create new salary DataFrame based on Salary z score
new_salary_df = df[(df['Salary_zscore']< 3) & (df['Salary_zscore'] > -3)]
new_salary_df

Company Name                  Job Title        Role  \
0                    Axis Bank             Data Scientist  Specialist   
1     Globex Digital Solutions  Machine Learning Engineer     Manager   
2           Tempworks Services             Data Scientist     Manager   
3                   BlueOptima  Machine Learning Engineer  Specialist   
4            Dell Technologies               Data Analyst  Specialist   
...                        ...                        ...         ...   
5122                   Rejolut  Machine Learning Engineer  Specialist   
5123     Bartleby Technologies  Machine Learning Engineer  Specialist   
5124                Clairvolex             Data Scientist     Manager   
5125       Piramal Enterprises               Data Analyst     Manager   
5126                  Unilever             Data Scientist  Specialist   

      Salaries Reported   Location     Salary  Salary_zscore  
0                     3  Bangalore 1171687.00          -0.03  
1                     1  Hyderabad 3257305.00           1.11  
2                     1  Bangalore 5811205.00           2.52  
3                     2  Bangalore 1180965.00          -0.03  
4                     5  New Delhi  602930.00          -0.35  
...                 ...        ...        ...            ...  
5122                  1     Mumbai  513690.00          -0.40  
5123                  1  Bangalore 1309535.00           0.04  
5124                  1  Bangalore 4499222.00           1.80  
5125                  1     Mumbai 4612711.00           1.86  
5126                  2  Bangalore 1705561.00           0.26  

[5064 rows x 7 columns]

<h4>Z score is one of standard measurements used to measure how far data point is differing from its average value, z-score of 3 means that data point is three deviations. These are outliers, which should be deleted. When setting z-score treshold to 3, mean will remove any data point that is three or more deviations away from average value(mean). This way we can eliminate extreme deviations in data. Let's see how many outliers we will remove from this data-analysis.<h4>

In [ ]:
# Factorize the Role column & add the new column into Data
label1, unique1 = pd.factorize(df['Role'], sort=False)
df['ManagerRole'] = label1
df

Company Name                  Job Title        Role  \
0                    Axis Bank             Data Scientist  Specialist   
1     Globex Digital Solutions  Machine Learning Engineer     Manager   
2           Tempworks Services             Data Scientist     Manager   
3                   BlueOptima  Machine Learning Engineer  Specialist   
4            Dell Technologies               Data Analyst  Specialist   
...                        ...                        ...         ...   
5122                   Rejolut  Machine Learning Engineer  Specialist   
5123     Bartleby Technologies  Machine Learning Engineer  Specialist   
5124                Clairvolex             Data Scientist     Manager   
5125       Piramal Enterprises               Data Analyst     Manager   
5126                  Unilever             Data Scientist  Specialist   

      Salaries Reported   Location     Salary  Salary_zscore  ManagerRole  
0                     3  Bangalore 1171687.00          -0.03            0  
1                     1  Hyderabad 3257305.00           1.11            1  
2                     1  Bangalore 5811205.00           2.52            1  
3                     2  Bangalore 1180965.00          -0.03            0  
4                     5  New Delhi  602930.00          -0.35            0  
...                 ...        ...        ...            ...          ...  
5122                  1     Mumbai  513690.00          -0.40            0  
5123                  1  Bangalore 1309535.00           0.04            0  
5124                  1  Bangalore 4499222.00           1.80            1  
5125                  1     Mumbai 4612711.00           1.86            1  
5126                  2  Bangalore 1705561.00           0.26            0  

[5123 rows x 8 columns]

<h3>Total of 63 outliers were removed from the analysis of the dataset.<h3>

In [ ]:
# Calculate Spearman correlations
corr_managerrole_salary = df['ManagerRole'].corr(df['Salary'], method='spearman')
corr_managerrole_sr = df['ManagerRole'].corr(df['Salaries Reported'], method='spearman')

# Display calculated Spearman correlations
print("Correlation between Manager Role and Salary:", corr_managerrole_salary)
print("Correlation between Manager Role and Salaries Reported:", corr_managerrole_sr)

Correlation between Manager Role and Salary: 0.4301196078482581
Correlation between Manager Role and Salaries Reported: -0.2544565723460406


<h4> A positive correlation of 0.4301196078482581 between 'ManagerRole' and 'Salary' columns suggests a moderate correlation between the two columns. This means that while the numerical label representing 'ManagerRole' increases, there is a tendency for the salary to increase. However, the correlation is that strong (over 0.5), indicating that the relationship may not be linear or that other factors would influence salary.

On the other hand, a correlation of -0.2544565723460406 between 'ManagerRole' and 'Salaries Reported' columns suggests a weak negative correlation. This means that as the numerical label representing the 'ManagerRole' increases, there is a slight tendency for the reported salaries to decrease. Again, the correlation is not that strong (-0.5), this indicates that the relationship may not be linear or that other factors would influence on reported salaries.
</h4>